### PR TITLE
Warn if main exists but is not exported

### PR DESCRIPTION
**Pull request by kripken**
As it may be dce'd, and then not run.

A possible reason this might happen is doing `./emcc -s EXPORTED_FUNCTIONS=["_myfunc"]` and forgetting to also have `_main` (which it is by default, but the commandline flag overrode the default value).

A possible downside to this patch is that it depends on emcc scanning symbols, which we have been hoping to avoid eventually. If we disallow that then logic like this would need to be in wasm-ld I guess.

### CODE REVIEW

File: tools/system_libs.py [hunk 1]
```diff
@@ -1105,6 +1105,16 @@ class libasan_rt_wasm(SanitizerLibrary):
   src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'asan']
 
 
+# If main() is not in EXPORTED_FUNCTIONS, it may be dce'd out. This can be
+# confusing, so issue a warning.
+def warn_on_unexported_main(symbolses):
+  if '_main' not in shared.Settings.EXPORTED_FUNCTIONS:
+    for symbols in symbolses:
+      if 'main' in symbolses[0].defs:
```
Comment (quantum5) [inline tools/system_libs.py]: `symbolses[0].defs`? Not `symbols.defs`?

(also I can't help but hear `symbolses` in a Gollum voice...)

Comment (kripken) [reply tools/system_libs.py]: Ah thanks, that was a copy-paste error...

Yes, it's an intentional lotr reference ;)

**Comment by Zireael07** [comment]
I just ran into this very issue today - absolutely no warning, just the project freezing.

BTW the same thing occurred with EMSCRIPTEN_KEEPALIVE macro.

**Comment by sbc100** [comment]
This feature is know to work in general it is working so I'm curing how you didn't see this warning:

```
$ emcc ~/test/hello.c -sEXPORTED_FUNCTIONS="[foo]"
system_libs:WARNING: main() is in the input files, but "_main" is not in EXPORTED_FUNCTIONS, which means it may be eliminated as dead code. Export it if you want main() to run.
```

@Zireael07 can you explain what you mean by "the same thing occurred with EMSCRIPTEN_KEEPALIVE macro."   What problem occurs with this macro?

**Comment by Zireael07** [comment]
The same problem occurred, project freezing. I will double check the build logs (it builds via GitHub Actions), but I am nearly 100% sure I didn't see a warning to that effect.

**Comment by sbc100** [comment]
And your problem is resolved by adding `_main` to your `EXPORTED_FUNCTIONS`

**Comment by Zireael07** [comment]
Yes, it was resolved. I suspect GitHub Actions might have suppressed the warning output somehow... At any rate, this is now solved.

**Pull request merged**